python/cudf_polars/cudf_polars/dsl/expr.py

```diff
@@ -278,54 +278,84 @@ def collect_agg(self, *, depth: int) -> AggInfo:
         )
 
 
-class NamedExpr(Expr):
-    __slots__ = ("name", "children")
-    _non_child = ("dtype", "name")
+class NamedExpr:
+    __slots__ = ("name", "value")
 
-    def __init__(self, dtype: plc.DataType, name: str, value: Expr) -> None:
-        super().__init__(dtype)
+    def __init__(self, name: str, value: Expr) -> None:
         self.name = name
-        self.children = (value,)
+        self.value = value
+
+    def __hash__(self) -> int:
+        """Hash of the expression."""
+        return hash((type(self), self.name, self.value))
+
+    def __repr__(self) -> str:
+        """Repr of the expression."""
+        return f"NamedExpr({self.name}, {self.value}"
+
+    def __eq__(self, other) -> bool:
+        """Equality of two expressions."""
+        return (
+            type(self) is type(other)
+            and self.name == other.name
+            and self.value == other.value
+        )
 
-    def do_evaluate(
+    def __ne__(self, other) -> bool:
+        """Inequality of expressions."""
+        return not self.__eq__(other)
+
+    def evaluate(
         self,
         df: DataFrame,
         *,
         context: ExecutionContext = ExecutionContext.FRAME,
-        mapping: dict[Expr, Column] | None = None,
-    ) -> Column:
+        mapping: Mapping[Expr, Column] | None = None,
+    ) -> NamedColumn:
         """Evaluate this expression given a dataframe for context."""
-        (child,) = self.children
-        return Column(
-            child.evaluate(df, context=context, mapping=mapping).obj, self.name
-        )
+        obj = self.value.evaluate(df, context=context, mapping=mapping)
+        if isinstance(obj, Scalar):
+            return NamedColumn(
+                plc.Column.from_scalar(obj.obj, 1),
+                self.name,
+                is_sorted=plc.types.Sorted.YES,
+                order=plc.types.Order.ASCENDING,
+                null_order=plc.types.NullOrder.BEFORE,
+            )
+        else:
+            return NamedColumn(
+                obj.obj,
+                self.name,
+                is_sorted=obj.is_sorted,
+                order=obj.order,
+                null_order=obj.null_order,
+            )
 
     def collect_agg(self, *, depth: int) -> AggInfo:
         """Collect information about aggregations in groupbys."""
-        (value,) = self.children
-        return value.collect_agg(depth=depth)
+        return self.value.collect_agg(depth=depth)
 
 
 class Literal(Expr):
     __slots__ = ("value",)
     _non_child = ("dtype", "value")
     value: pa.Scalar
 
-    def __init__(self, dtype: plc.DataType, value: Any) -> None:
+    def __init__(self, dtype: plc.DataType, value: pa.Scalar) -> None:
         super().__init__(dtype)
-        self.value = pa.scalar(value)
+        assert value.type == plc.interop.to_arrow(dtype)
```

Comment:
Does the code depend on this for correct execution? Python can be run with `-O` which disables all assertions. If this is necessary for correctness (i.e. if this library or user code would be expected to catch and handle an `AssertionError`), we should instead raise a `ValueError` explicitly if the condition is false.

Reply:
This is belt-and-braces, the only construction of this object is by a caller who guarantees that the `value` was made with the dtype constructed from `plc.interop.to_arrow(dtype)`